packages/ui5-application-writer/src/data/defaults.ts

```diff
@@ -110,20 +111,37 @@ export function mergeUi5(ui5: Partial<UI5>): UI5 {
         framework,
         frameworkUrl: ui5.frameworkUrl ?? defaultFrameworkUrl
     };
-    // typesVersion must be a valid npm semantic version, we know they cannot be null as already validated
-    const localSemVer = semVer.valid(semVer.coerce(merged.localVersion))!;
-    const typesVersion = semVer.gte(localSemVer, UI5_DEFAULT.TYPES_VERSION_SINCE)
-        ? localSemVer
-        : UI5_DEFAULT.TYPES_VERSION_PREVIOUS;
 
     merged.descriptorVersion = getManifestVersion(merged.minUI5Version, ui5.descriptorVersion);
-    merged.typesVersion = ui5.typesVersion ?? typesVersion;
+    merged.typesVersion = ui5.typesVersion ?? getTypesVersion(merged.minUI5Version);
     merged.ui5Theme = ui5.ui5Theme ?? 'sap_fiori_3';
     merged.ui5Libs = getUI5Libs(ui5.ui5Libs);
 
     return Object.assign({}, ui5, merged) as UI5;
 }
 
+/**
+ * Get the best types version for the given minUI5Version.
+ * For the latest versions the lates S/4 on-premise version (1.102.x) is used, for anything before we match the versions as far back as available.
+ *
+ * @param minUI5Version the mininum UI5 version that needs to be supported
+ * @returns semantic version representing the types version.
+ */
+export function getTypesVersion(minUI5Version?: string) {
+    const version = semVer.coerce(minUI5Version);
+    if (!version) {
+        return `~${UI5_DEFAULT.TYPES_VERSION_BEST}`;
+    } else {
+        if (semVer.gte(version, UI5_DEFAULT.TYPES_VERSION_BEST)) {
+            return `~${UI5_DEFAULT.TYPES_VERSION_BEST}`;
+        } else {
```

Comment:
What about if the version is greater than 1.108.0. Shouldn't that version be used rather than 1.102.x?

Reply:
the idea was that if in 1.109 something is missing, then we also downport it to the `BEST` version, so for the types, the `BEST` version should be your best shot.

Reply:
OK. Interesting. Fine for now but I assume that logic will need to reviewed at some point